substrate/network-libp2p/src/service.rs

```diff
@@ -0,0 +1,1106 @@
+		swarm_controller);
+
+	// Merge all the futures into one!
+	Ok(swarm_future
+		.select(discovery).map_err(|(err, _)| err).and_then(|(_, rest)| rest)
+		.select(timeouts).map_err(|(err, _)| err).and_then(|(_, rest)| rest)
+		.select(close_rx.then(|_| Ok(()))).map(|_| ()).map_err(|(err, _)| err)
+
+		.and_then(move |_| {
+			debug!(target: "sub-libp2p", "Networking ended ; disconnecting \
+				all peers");
+			shared.network_state.disconnect_all();
+			Ok(())
+		}))
+}
+
+/// Output of the common transport layer.
+struct TransportOutput<S> {
+	socket: S,
+	info: Box<Future<Item = IdentifyTransportOutcome, Error = IoError>>,
+	original_addr: Multiaddr,
+}
+
+/// Enum of all the possible protocols our service handles.
+enum FinalUpgrade<C> {
+	Kad((KadConnecController, Box<Stream<Item = KadIncomingRequest, Error = IoError>>)),
+	/// The remote identification system, and the multiaddress we see the remote as.
+	Identify(IdentifyOutput<C>, Multiaddr),
+	Ping(ping::Pinger, Box<Future<Item = (), Error = IoError>>),
+	/// `Custom` means anything not in the core libp2p and is handled
+	/// by `CustomProtoConnectionUpgrade`.
+	Custom(RegisteredProtocolOutput<Arc<NetworkProtocolHandler + Send + Sync>>),
+}
+
+/// Called whenever we successfully open a multistream with a remote.
+fn listener_handle<'a, C>(
+	shared: Arc<Shared>,
+	upgrade: FinalUpgrade<C>,
+	endpoint: Endpoint,
+	client_addr: impl Future<Item = Multiaddr, Error = IoError> + 'a,
+	/*listener_upgrade: impl ConnectionUpgrade<C, Box<Future<Item = Multiaddr, Error = IoError>>>*/
+) -> Box<Future<Item = (), Error = IoError> + 'a>
+	where C: AsyncRead + AsyncWrite + 'a {
+	match upgrade {
+		FinalUpgrade::Kad((controller, kademlia_stream)) => {
+			trace!(target: "sub-libp2p", "Opened kademlia substream with \
+				remote as {:?}", endpoint);
+
+			let shared = shared.clone();
+			Box::new(client_addr.and_then(move |client_addr|
+				handle_kademlia_connection(shared, client_addr, controller, kademlia_stream)
+			).flatten())
+		},
+
+		FinalUpgrade::Identify(IdentifyOutput::Sender { sender }, original_addr) => {
+			trace!(target: "sub-libp2p", "Sending back identification info");
+			/*let protocols_to_report = ConnectionUpgrade::<_, Box<Future<Item = Multiaddr, Error = IoError>>>::protocol_names(&listener_upgrade)
+				.filter_map(|(name, _)| String::from_utf8(name.to_vec()).ok())
+				.collect();*/
+			sender.send(
+				IdentifyInfo {
+					public_key: shared.network_state.local_public_key().clone(),
+					protocol_version: concat!("substrate/",
+						env!("CARGO_PKG_VERSION")).to_owned(),		// TODO: ?
+					agent_version: concat!("substrate/",
+						env!("CARGO_PKG_VERSION")).to_owned(),
+					listen_addrs: shared.listened_addrs.read().clone(),
+					protocols: Vec::new(),		// TODO: protocols_to_report,
+				},
+				&original_addr
+			)
+		},
+
+		FinalUpgrade::Identify(IdentifyOutput::RemoteInfo { .. }, _) =>
+			unreachable!("We are never dialing with the identify protocol"),
```

Comment:
what if someone dials us?

Reply:
If someone dials us, we enter the `FinalUpgrade::Identify(IdentifyOutput::Sender { sender }, original_addr)` block.